src/ol/layer/Heatmap.js

```diff
@@ -285,19 +285,23 @@ class Heatmap extends BaseVector {
 
             uniform sampler2D u_image;
             uniform sampler2D u_gradientTexture;
+            uniform float u_opacity;
 
             varying vec2 v_texCoord;
 
             void main() {
               vec4 color = texture2D(u_image, v_texCoord);
-              gl_FragColor.a = color.a;
+              gl_FragColor.a = color.a * u_opacity;
               gl_FragColor.rgb = texture2D(u_gradientTexture, vec2(0.5, color.a)).rgb;
```

Comment:
Should we be sampling the gradient texture after applying the layer opacity or before?  Feels to me like it should be after, but I'm not looking at the results.

Reply:
You can test in https://codesandbox.io/s/simple-forked-yrj3i?file=/main.js  Doing it after makes 50% opacity look much less than I would expect.

Reply:
Layer opacity should definitely be applied after the gradient texture sampling. Here the `color.a` value isn't meant to be an opacity but more of an intensity value, which can then be mapped to a color.

Reply:
This makes sense as it is actually, the layer opacity is only applied to `gl_FragColor` and not `color.a` so it works well.